### PR TITLE
fix: windows gen file

### LINKED ISSUE #955: dev-scripts path issue on Windows
**Describe the bug**
Running `npm run gen` throws the following error on Windows:

image

**To Reproduce**
- Be on Windows 10
- Run `npm install`
- Run `npm run gen` from the root of the project or from examples dir
- See the same error as the screenshot above

**Misc**
- Node version: 18.20.3
- Package manager: pnpm
- Browser: Chrome
- [ ] I'm a [sponsor](https://github.com/sponsors/YousefED) and would appreciate if you could look into this sooner than later 💖

---

I was hitting my head for an hour trying to figure this one out. When I comment out `fs.writeFileSync` the error is gone, and this seems to be originating from the way we're getting the path/dir :

```
const dir = path.parse(import.meta.url.replace("file://", "")).dir;
```

This issue is in every instance of the above code, in `gen.ts`, `genDocs.ts`, and `util.ts`

---
I was only able to generate the example code in #954 using my MacOS. So this issue is exclusively on Windows.

### PR DESCRIPTION
**Pull request by jkcs**
/claim #955
close #955

### CODE REVIEW

**Comment by zaaakher** [comment]
Good progress 👍 

I just tested it on both Windows 10 and MacOS and here are the current results


| OS      | Updated `examples.gen.tsx` | adds example files |
| ------- | ------------------------ | ------------------ |
| Windows | ✅                       | ❌                 |
| MacOS   | ✅                       | ✅                 |



By "adds example files" I mean the `package.json`, `tsconfig.json`, `index.html`, `main.tsx`, and `vite.config.ts` in each example that should be automatically generated by the script.

**Comment by zaaakher** [comment]
@jkcs I think the only part missing is in the `writeTemplate` function in `gen.ts` (line 27)

**Comment by zaaakher** [comment]
I just tested it again and everything seems to be working as expected in both Windows 10 and MacOS 👍 

| OS      | Updates `examples.gen.tsx` | Adds example files |
| ------- | ------------------------ | ------------------ |
| Windows | ✅                       | ✅                 |
| MacOS   | ✅                       | ✅                 |

@jkcs Thanks for your work 👍

File: packages/dev-scripts/examples/gen.ts [hunk 1]
```diff
@@ -57,7 +58,7 @@ async function writeTemplate(project: Project, templateFile: string) {
 
 async function generateCodeForExample(project: Project) {
   const templates = glob.sync(
-    path.resolve(dir, "./template-react/*.template.tsx")
+    path.resolve(dir, "./template-react/*.template.tsx").replace(/\\/g, "/")
```
Comment (YousefED) [inline packages/dev-scripts/examples/gen.ts:61]: instead of replace, using `path.join` is cleaner?

File: packages/dev-scripts/examples/util.ts [hunk 2]
```diff
@@ -114,7 +116,7 @@ export type Files = Record<
 
 export function getProjectFiles(project: Project): Files {
   const dir = path.resolve("../../", project.pathFromRoot);
-  const files = glob.globSync(dir + "/**/*", {
+  const files = glob.globSync(glob.win32.convertPathToPattern(dir + "/**/*"), {
```
Comment (YousefED) [inline packages/dev-scripts/examples/util.ts:119]: Now we're also calling win32.convertPathToPattern on mac / unix right? Does that make sense? It seems a bit weird to me!

File: packages/dev-scripts/examples/util.ts [hunk 3]
```diff
@@ -140,7 +142,11 @@ export function getProjectFiles(project: Project): Files {
  */
 export function getExampleProjects(): Project[] {
   const examples: Project[] = glob
-    .globSync(path.join(dir, "../../../examples/**/*/.bnexample.json"))
+    .globSync(
+      glob.win32.convertPathToPattern(
```
Comment (YousefED) [inline packages/dev-scripts/examples/util.ts:146]: same question as above

File: packages/dev-scripts/examples/util.ts [hunk 4]
```diff
@@ -162,9 +168,8 @@ export function getExampleProjects(): Project[] {
         .split(path.sep);
 
       const group = {
-        pathFromRoot: path.relative(
-          path.resolve("../../"),
-          path.join(directory, "..")
+        pathFromRoot: slash(
```
Comment (YousefED) [inline packages/dev-scripts/examples/util.ts:171]: I'm not sure `slash` is needed (would prefer not to add this dependency). maybe we can just prevent passing / to path.resolve? e.g.: `path.resolve("..","..")`

**Review by YousefED — COMMENTED** [review]
great! some questions

File: packages/dev-scripts/examples/gen.ts (hunk 1 above)
Comment (YousefED) [reply packages/dev-scripts/examples/gen.ts:61]: ah, and the is that `glob` always expects a fwd slash I suppose?

File: packages/dev-scripts/examples/util.ts (hunk 4 above)
Comment (YousefED) [reply packages/dev-scripts/examples/util.ts:171]: Yes, I'd love to see this code a bit clearer / more uniform. When reviewing, it's not entirely clear to me why we have 3 different methods:
- `replace`
- `slash`
- `convertPathToPattern`

Can we either use 1 method (ideally with a separate, well named helper function). Or, if multiple methods are necessary, add a comment in code why we use each function?

File: packages/dev-scripts/examples/util.ts [hunk 5]
```diff
@@ -201,3 +202,13 @@ export function getExampleProjects(): Project[] {
   // });
   return examples;
 }
+
+export function replacePathSepToSlash(path: string) {
+  const isExtendedLengthPath = path.startsWith("\\\\?\\");
```
Comment (YousefED) [inline packages/dev-scripts/examples/util.ts:207]: what's an example of this?

**Review by YousefED — APPROVED** [review]
Thanks @jkcs for your contribution and collaboration and @zaaakher for testing :)

**Pull request merged**